File: doublebeam/raytracing/twopoint.py
import math
from math import sqrt, sin, cos

import numpy as np

from doublebeam.models import VelocityModel3D
from doublebeam.utils import Index, safe_divide, horizontal_distance, angle_clockwise


class TwoPointRayTracing:
    """
    Implement two point ray tracing as described in
    A fast and robust two-point ray tracing method in layered media with
    constant or linearly varying layer velocity (Xinding Fang and Xiaofei Chen
    2019)
    This implementation only works for direct rays between the source and
    receiver. Turning rays and reflected rays are not implemented.
    """

    def __init__(self, velocity_model: VelocityModel3D):
        """
        Initialize two point ray tracing for the given velocity model
        """
        self._model = velocity_model
        self._a = velocity_model.gradients
        self._b = velocity_model.intercepts
        self._n = len(velocity_model)
        self._z = velocity_model.interface_depths

    def _v_M(self, source_position: np.ndarray,
             receiver_position: np.ndarray) -> float:
        """
        Return the highest velocity of the layers the direct ray passes through
        on the way from source to receiver.
        :param source_position: x, y, z triple of source coordinates
        :param receiver_position: x, y, z triple of receiver coordinates
        :return: Highest velocity (m/s) between source and receiver for a direct
        ray
        """
        receiver_index = self._model.layer_index(receiver_position)
        source_index = self._model.layer_index(source_position)
        if source_index == receiver_index:
            return max(self._model.eval_at(source_position),
                       self._model.eval_at(receiver_position))
        # first, get all the top and bottom velocities from the interfaces
        # between the source and receiver
        # sort indices because accessing array only works from low to high
        index_low = min(source_index, receiver_index)
        index_high = max(source_index, receiver_index)
        # dont +1 last index since the bottom of the bottom most layer is not
        # passed by the ray. If the ray starts there, evaluating the velocity
        # at source and receiver will catch it.
        velocities_bottom = self._model.velocities_bot[index_low:index_high]
        # +1 for first index to exclude top velocity of the layer containing the
        # upper point.
        # There are two cases:
        # - The point lays below the top of the layer, so the ray doesn't pass
        #   through the velocity at the top of the layer
        # - The point lays on the top depth, so the velocity should be included.
        #   This case in then handled by evaluating the velocity at both points.
        # +1 last index to include top velocity of the bottom most layer
        velocities_top = self._model.velocities_top[index_low+1:index_high+1]
        # second, also get velocity at the end points (source and receiver)
        v = max(self._model.eval_at(source_position),
                self._model.eval_at(receiver_position))
        return max(np.max(velocities_bottom), np.max(velocities_top), v)

    def trace(self, source: np.ndarray, receiver: np.ndarray,
              accuracy: float = 1E-10) -> np.ndarray:
        """
        Two point ray tracing from source to receiver position
        :param source: array of shape (3,) which contains the x y z coordinates
        of the source point
        :param receiver: array of shape (3,) which contains the x y z
        coordinates of the receiver point
        :param accuracy: Iteration for modified ray parameter q will be stopped
        if the difference between the current and the next value falls below
        this value.
        :return: Ray parameters px, py, pz in s/m
        """

        def initial_q():
            alpha1 = d1
            alpha2 = c0 * (c0**2 + d1 * cminus1) / (cminus1**2 - c0 * cminus2)
            beta1 = (c0 * cminus1 + d1 * cminus2) / (c0 * cminus2 - cminus1**2)
            beta2 = (c0**2 + d1 * cminus1) / (cminus1**2 - c0 * cminus2)
            numerator = (beta1 * X - alpha1 + sqrt((beta1**2 - 4 * beta2) * X**2
                                                   + 2 * (2 * alpha2 - alpha1 * beta1) * X
                                                   + alpha1**2))
            denominator = 2 * (alpha2 - beta2 * X)
            return numerator / denominator

        def X_tilde(q):
            return delta_a * mu_tilde * (np.sqrt(q**-2 + epsilon_tilde) - np.sqrt(q**-2 + omega_tilde)) \
                   + (1 - delta_a) * h_tilde / np.sqrt(q**-2 + epsilon_tilde)

        def X_tilde_prime(q):
            return delta_a * mu_tilde / (q**2) * (1 / np.sqrt(1 + omega_tilde * q**2)
                                                  - 1 / np.sqrt(1 + epsilon_tilde * q**2)) \
                   + (1 - delta_a) * h_tilde / (1 + epsilon_tilde * q**2)**1.5

        def X_tilde_double_prime(q):
            return delta_a * mu_tilde / q**3 * ((2 + 3 * epsilon_tilde * q**2) / (1 + epsilon_tilde * q**2)**1.5
                                                - (2 + 3 * omega_tilde * q**2) / (1 + omega_tilde * q**2)**1.5) \
                   - (1 - delta_a) * 3 * h_tilde * epsilon_tilde * q / (1 + epsilon_tilde * q**2)**2.5

        def f_tilde(q):
            return np.sum(X_tilde(q)) - X

        def f_tilde_prime(q):
            return np.sum(X_tilde_prime(q))

        def f_tilde_double_prime(q):
            return np.sum(X_tilde_double_prime(q))

        def next_q(q_old):
            A = 0.5 * f_tilde_double_prime(q)
            B = f_tilde_prime(q)
            C = f_tilde(q)
            delta_q_plus = (-B + sqrt(B**2 - 4 * A * C)) / (2 * A)
            delta_q_minus = (-B - sqrt(B**2 - 4 * A * C)) / (2 * A)
            q_plus = q_old + delta_q_plus
            q_minus = q_old + delta_q_minus
            if abs(f_tilde(q_plus)) < abs(f_tilde(q_minus)):
                return q_plus
            else:
                return q_minus

        top, bottom = self._model.vertical_boundaries()
        if not top <= source[Index.Z] <= bottom:
            raise ValueError(f"Source {source} outside of model")
        if not top <= receiver[Index.Z] <= bottom:
            raise ValueError(f"Receiver {receiver} outside of model")

        zs = source[Index.Z]
        a = self._a
        b = self._b
        n = self._n
        z = self._z

        # because paper uses 1 based indexing
        s = self._model.layer_index(zs) + 1
        # special case: bottom of the lowest layer should belong to the lowest
        # layer instead of the layer below as for the other layer
        if s > n:
            s = n

        a = np.insert(a, 0, a[s - 1])
        b = np.insert(b, 0, b[s - 1])

        # eq. A5
        epsilon = np.array([(a[k] * z[k - 1] + b[k])**2 for k in range(1, n + 1)])
        epsilon = np.insert(epsilon, 0, (a[s] * z[s - 1] + b[0])**2)

        # eq. A6
        omega = np.array([(a[k] * z[k] + b[k])**2 for k in range(1, n + 1)])
        omega = np.insert(omega, 0, (a[s] * zs + b[s])**2)

        # eq. A7
        h = np.array([(a[k] * z[k - 1] + b[k]) * (z[k] - z[k - 1]) for k in range(1, n + 1)])
        h = np.insert(h, 0, (a[s] * z[s - 1] + b[s]) * (zs - z[s - 1]))

        source_below_receiver = source[Index.Z] > receiver[Index.Z]
        mu = np.array([_mu_k(k, s, n, source_below_receiver) for k in range(0, n + 1)])
        vM = self._v_M(source, receiver)

        # eq. A10
        mu_tilde = np.divide(mu * vM, a, out=np.zeros_like(a), where=a != 0)

        # eq. A11
        h_tilde = mu * h / vM

        # eq. A12
        epsilon_tilde = 1 - epsilon / vM**2

        # eq. A13
        omega_tilde = 1 - omega / vM**2

        # eq.A9
        delta_a = np.abs(np.sign(a))

        # eq. C13
        d1 = np.sum(delta_a * 0.5 * mu_tilde * (epsilon_tilde - omega_tilde) + (1 - delta_a) * h_tilde)

        # eq. C18
        delta_epsilon = np.abs(np.sign(epsilon_tilde))

        # eq. C19
        delta_omega = np.abs(np.sign(omega_tilde))

        #eq. C14
        c0 = (delta_a * mu_tilde * (delta_epsilon * np.sqrt(epsilon_tilde) - delta_omega * np.sqrt(omega_tilde))
              + np.divide((1 - delta_a) * delta_epsilon * h_tilde, np.sqrt(epsilon_tilde), out=np.zeros_like(a),
                          where=epsilon_tilde != 0))
        c0 = np.sum(c0)

        # eq. C16
        cminus1 = np.sum(delta_a * (delta_omega - delta_epsilon) * mu_tilde)

        # eq. C17
        cminus2 = (delta_a * 0.5 * mu_tilde * (
                safe_divide(delta_epsilon, np.sqrt(epsilon_tilde)) - safe_divide(delta_omega, np.sqrt(omega_tilde)))
                   - safe_divide((1 - delta_a) * delta_epsilon * h_tilde, (2 * epsilon_tilde**1.5)))
        cminus2 = np.sum(cminus2)

        # horizontal distance between source and receiver
        X = horizontal_distance(source, receiver)
        q = initial_q()
        while True:
            q_next = next_q(q)
            if abs(q - q_next) < accuracy:
                q = q_next
                break
            q = q_next

        horizontal_slowness = _q_to_p(q, vM)
        c = self._model.eval_at(source)
        vertical_slowness = math.sqrt(c**-2 - horizontal_slowness**2)
        if source_below_receiver:
            # ray should travel upward from source in this case
            vertical_slowness *= -1
        receiver_surface_projection = receiver - source
        receiver_surface_projection[Index.Z] = 0
        x_axis = np.array((1, 0, 0))
        phi = angle_clockwise(x_axis, receiver_surface_projection)
        px = cos(phi) * horizontal_slowness
        py = sin(phi) * horizontal_slowness
        return np.array((px, py, vertical_slowness))


def _mu_k(k, s, n, source_below_receiver: bool = True) -> int:
    """
    Eq. A8
    :param k: Index of current layer
    :param s: Index of source layer
    :param n: Number of layers
    :param source_below_receiver: Fang2019s algorithm is only designed for
    sources below the receiver. If the source is above the receiver, this
    function has to be modified. Mu_k normally returns 1 for all layers above
    the source layer and 0 for all layers below. This must be reversed if the
    source is above the receiver
    """
    if 1 <= k <= s - 1:
        return 1 if source_below_receiver else 0
    if s <= k <= n:
        return 0 if source_below_receiver else 1
    if k == 0:
        return 1 - _mu_k(s, s, n, source_below_receiver)


def _q_to_p(q: float, v_M: float) -> float:
    """
    Backtransform modified ray parameter q to standard ray parameter p.
    Reordered from eq. 15.
    :param q: modified ray parameter
    """
    return sqrt(q**2 / (v_M**2 + v_M**2 * q**2))

File: doublebeam-cpp/src/twopoint.cpp
#include "twopoint.hpp"
#include "utils.hpp"
#include <cmath>
#include <iostream>
#include <iterator>
#include <valarray>


template <typename T>
std::ostream& operator<<(std::ostream& os, const std::valarray<T>& array) {
    os << "(";
    for (auto item : array) {
        os << item << " ";
    }
    os << ")";
    return os;
}

#define USEDEBUG false
#if USEDEBUG
#define msg(x) std::cout << #x << " = " << x << std::endl
#else
#define msg(x)
#endif

/**
 * Sum values in array treating nan as zero.
 * @param in
 * @return
 */
TwoPointRayTracing::array_t::value_type nansum(const TwoPointRayTracing::array_t& in) {
    TwoPointRayTracing::array_t::value_type sum{0};
    for (auto e : in) {
        if (std::isnan(e)) {
            continue;
        }
        sum += e;
    }
    return sum;
}


std::string stringify(position_t pos) {
    auto [x, y, z] = pos;
    return "(" + std::to_string(x) + " " + std::to_string(y) + " " + std::to_string(z) + ")";
}

// eq. A3
template <typename A>
TwoPointRayTracing::array_t X_tilde(double q, const A& epsilon_tilde, const A& h_tilde) {
    return h_tilde / std::sqrt(1 / (q * q) + epsilon_tilde);
}

// eq. B9
template <typename A>
TwoPointRayTracing::array_t X_tilde_prime(double q, const A& epsilon_tilde, const A& h_tilde) {
    return h_tilde / std::pow(1 + epsilon_tilde * q * q, 1.5);
}

// eq. B10
template <typename A>
TwoPointRayTracing::array_t X_tilde_double_prime(double q, const A& epsilon_tilde,
                                                 const A& h_tilde) {
    return -3 * h_tilde * epsilon_tilde * q / std::pow(1 + epsilon_tilde * q * q, 2.5);
}


// eq. 16
template <typename A>
double f_tilde(double q, double X, const A& epsilon_tilde, const A& h_tilde) {
    return nansum(X_tilde(q, epsilon_tilde, h_tilde)) - X;
}

// eq. B3
template <typename A>
double f_tilde_prime(double q, const A& epsilon_tilde, const A& h_tilde) {
    return nansum(X_tilde_prime(q, epsilon_tilde, h_tilde));
}

template <typename A>
double f_tilde_double_prime(double q, const A& epsilon_tilde, const A& h_tilde) {
    return nansum(X_tilde_double_prime(q, epsilon_tilde, h_tilde));
}

// Solve nonlinear equation (5) using Newton algorithm with cubic iteration.
template <typename A>
double next_q(double q, double X, const A& epsilon_tilde, const A& h_tilde) {
    double f_double_prime = f_tilde_double_prime(q, epsilon_tilde, h_tilde);
    double f_prime = f_tilde_prime(q, epsilon_tilde, h_tilde);
    double f = f_tilde(q, X, epsilon_tilde, h_tilde);
    // use approximation which is valid for small a.
    // 1-sqrt(1-a) = a/2 + a^2/8 + O(a^3)
    // see http://numbers.computation.free.fr/Constants/Algorithms/newton.html
    auto q_new = q - f / f_prime * (1 + (f * f_double_prime / (2 * f_prime * f_prime)));
    return q_new;
}

// Transformed version of eq. (15) for p
double q_to_p(double q, double vM) {
    return std::sqrt(q * q / (vM * vM + vM * vM * q * q));
}


template <typename T>
T Heaviside(T x) {
    return x < 0 ? T{0.} : T{1};
}


double safe_divide(double numerator, double denominator) {
    if (numerator == 0) {
        return numerator;
    }
    return numerator / denominator;
}


// Calculate starting value for q using either C1 or C6 depending on value of c1.
// For constant velocity layers c_minus1 will always be 0, so it is removed from the formulas.
double initial_q(double X, double d1, double c0, double c1, double cminus2) {
    auto alpha1 = d1;
    msg(alpha1);
    auto alpha2 = safe_divide(c0 * c0 * c0, -c0 * cminus2);
    msg(alpha2);
    auto beta1 = safe_divide(d1 * cminus2, c0 * cminus2);
    msg(beta1);
    auto beta2 = safe_divide(c0 * c0, -c0 * cminus2);
    msg(beta2);

    if (c1 == 0) {
        // Use equation C1 to estimate initial value for q
        auto numerator = (beta1 * X - alpha1 +
                          sqrt((beta1 * beta1 - 4 * beta2) * X * X +
                               2 * (2 * alpha2 - alpha1 * beta1) * X + alpha1 * alpha1));
        msg(numerator);
        auto denominator = 2 * (alpha2 - beta2 * X);
        msg(denominator);
        double q = numerator / denominator;
        return q;
    } else {
        // This uses equation C6 to estimate initial value for q. The paper mistakenly states this
        // equation should be used when c_1 = 0, but it should be used when c1 != 0. This was
        // confirmed by email from the corresponding author.
        auto q = X / alpha1 + (X / c1 - X / alpha1 - c0 / c1) *
                                  Heaviside(X - safe_divide(alpha1 * c0, alpha1 - c1));
        return q;
    }
}

double horizontal_distance(double x1, double y1, double x2, double y2) {
    return std::sqrt(std::pow(x1 - x2, 2) + std::pow(y1 - y2, 2));
}


TwoPointRayTracing::TwoPointRayTracing(VelocityModel velocity_model) :
        model(std::move(velocity_model)) {}


Slowness TwoPointRayTracing::trace(Position source, Position receiver, double accuracy,
                                   int max_iterations) {
    auto [min_depth, max_depth] = std::minmax(source.z.get(), receiver.z.get());
    if (not model.in_model(source.x, source.y, source.z)) {
        throw std::domain_error(impl::Formatter()
                                << "Source at " << source << " outside of model.");
    }
    if (not model.in_model(receiver.x, receiver.y, receiver.z)) {
        throw std::domain_error(impl::Formatter()
                                << "Receiver at " << receiver << "  outside of model.");
    }
    auto source_index = model.layer_index(source.z).value();
    auto receiver_index = model.layer_index(receiver.z).value();
    auto [min_index, max_index] = std::minmax(source_index, receiver_index);
    auto num_layers = model.number_of_interfaces_between(source.z, receiver.z) + 1;
    msg(source_index);
    msg(receiver_index);
    msg(num_layers);
    // Only the layers between source and receiver and the layers containing source and receiver
    // matter. All others are ignored. Furthermore the depth in z will be set to source/receiver
    // depth at the top and bottom.
    array_t z(num_layers + 1);
    array_t b(num_layers);
    z[0] = min_depth;
    z[num_layers] = max_depth;
    b[0] = model[min_index].velocity.get();
    for (size_t i = 1; i < num_layers; ++i) {
        z[i] = model[min_index + i - 1].bot_depth.get();
        b[i] = model[min_index + i].velocity.get();
    }
    msg(z);
    msg(b);
    array_t epsilon = b * b;
    msg(epsilon);
    array_t h(num_layers);
    if (num_layers == 1) {
        h = b * std::abs((source.z - receiver.z).get());
    } else {
        for (size_t i = 0; i < num_layers; ++i) {
            h[i] = b[i] * (z[i + 1] - z[i]);
        }
    }
    msg(h);
    double vM = b.max();
    msg(vM);
    array_t h_tilde = h / vM;
    msg(h_tilde);
    array_t epsilon_tilde = 1 - epsilon / (vM * vM);
    msg(epsilon_tilde);
    double d1 = h_tilde.sum();
    // TODO find better way for sum, nansum evaluates into temporary array
    double c0 = nansum((h_tilde / sqrt(epsilon_tilde))[epsilon_tilde != 0]);
    double c1 = nansum(h_tilde[epsilon_tilde == 0]);
    double cn2 = -0.5 * nansum((h_tilde / pow(epsilon_tilde, 1.5))[epsilon_tilde != 0]);
    msg(d1);
    msg(c0);
    msg(c1);
    msg(cn2);

    auto X =
        horizontal_distance(source.x.get(), source.y.get(), receiver.x.get(), receiver.y.get());
    auto q = initial_q(X, d1, c0, c1, cn2);
    for (; max_iterations > 0; --max_iterations) {
        double q_next = next_q(q, X, epsilon_tilde, h_tilde);
        if (std::abs(q - q_next) < accuracy) {
            q = q_next;
            break;
        }
        q = q_next;
    }
    double horizontal_slowness = q_to_p(q, vM);
    Velocity c = model.eval_at(source.x, source.y, source.z).value();
    InverseVelocity vertical_slowness(
        std::sqrt(std::pow(c.get(), -2) - horizontal_slowness * horizontal_slowness));
    bool source_below_receiver = source.z > receiver.z;
    if (source_below_receiver) {
        // ray should travel upward from source in this case
        vertical_slowness *= -1;
    }
    // calculate angle to x axis
    double phi =
        math::angle_clockwise((receiver.x - source.x).get(), (receiver.y - source.y).get(), 1., 0.);
    // horizontal slowness is combination of px and py, now we have to split it up into its parts
    InverseVelocity px(std::cos(phi) * horizontal_slowness);
    InverseVelocity py(std::sin(phi) * horizontal_slowness);
    return {px, py, vertical_slowness};
}
